src/main.c

```diff
@@ -1,69 +1,123 @@
+     */
+    int length = argc-1;
+    int *data = calloc(sizeof(int), argc);
+    /* Get user input. note: first argument is at argv[1] */
+    for (int i = 0; i < length; ++i) {
+        data[i] = atoi(argv[i+1]);
+    }
 
-    int data_length = sizeof(data) / sizeof(data[0]);
+    /**
+     * The original way this was done was bad for two reasons.
+     * First of all, if the list were empty, then there would only be one node,
+     * and it would be completely uninitialized.
+     *
+     * Furthermore, for any other size, there will be one extra element at the
+     * end of the list (since the loop creates a node every loop, and there's
+     * one at the beginning, there will always be n+1 nodes).
+     * In either case, the last node is uninitialized, which means that there's
+     * no `NULL` at the end of the list which is really bad.
+     *
+     * Instead, we will set up the list differently.
+     */
+    struct node *head, *current;
+    head = current = NULL;
 
-    //set up our list
-    while(i < data_length) {
-        current -> item = data[i]; //set data for current node
-        //make a new node and point to it
-        current -> pnext = malloc(sizeof(struct node));
-        //point current to that new node
-        current = current -> pnext;
-        i++;
+    /* set up our list */
+    for (int i = 0; i < length; ++i) {
+        /* allocate the current node */
+        if (!current) {
+            /* if we don't have a current node, allocate it */
+            current = head = malloc(sizeof(struct node));
+        } else {
+            /* else just add it to the end of the list */
+            current->pnext = malloc(sizeof(struct node));
+            current = current->pnext;
+        }
+        current->item = data[i]; //set data for current node
     }
 
     current = head;
     //print it
-    while(current -> pnext != NULL) {
-        printf("(%d)\n", current -> item);
-        current = current -> pnext;
+    /**
+     * The former loop condition `current->pnext != null` was incorrect.
+     * This was printing all the elements except for the last one.
+     * This worked fine with the above, since there was always one extra node,
+     * but now there isn't so we change the iteration to go over the whole list.
+     */
+    while(current != NULL) {
+        printf("(%d) ", current->item);
+        current = current->pnext;
     }
     printf("\n");//blank line for formatting
 
-    //remove twice
+
+
+    // Remove twice
     for(int i = 0; i < 2; i++) {
         struct node *old_head = head;
         head = head -> pnext;
         free(old_head);
     }
 
-    //print again
-    current = head;
-    while(current -> pnext != NULL) {
-        printf("(%d)\n", current -> item);
-        current = current -> pnext;
+    /* Just for reference, another way to print it: */
+    for (current = head; current != NULL; current = current->pnext) {
+        printf("(%d) ", current->item);
     }
+    puts("\n");
```

Comment:
difference between and when which is preferred for `puts()` and `printf()`?

Reply:
`printf` prints a formatted string, i.e. which contains things like "%d" and such.

It's a variadic function (which means it takes multiple arguments)

`puts` on the other hand just prints a string. Really if I wanted to be cool i'd use `putc` (put char) but either of the three work.

Really, whenever you do `printf("unformatted string");` the compiler will probably just turn it into a call to `puts`